File: 415.字符串相加.cpp
/*
 * @lc app=leetcode.cn id=415 lang=cpp
 *
 * [415] 字符串相加
 */

// @lc code=start
#include<string>
using namespace std;
class Solution {
public:
    string ret;
    void func(const string& n1, const string& n2, int idx, int carry = 0) {
        if(n1.size() <= idx && n2.size() <= idx && carry == 0) return;
        int tmp = (idx < n1.size()? n1[idx] - '0': 0) + (idx < n2.size()? n2[idx] - '0': 0) + carry;
        ret.push_back(tmp % 10 + '0');
        func(n1, n2, idx + 1, tmp / 10);
    }
    string addStrings(string num1, string num2) {
        reverse(num1.begin(), num1.end());
        reverse(num2.begin(), num2.end());
        ret.clear();
        func(num1, num2, 0);
        reverse(ret.begin(), ret.end());
        return ret;
    }
};
// @lc code=end

File: 415.字符串相加.py
#
# @lc app=leetcode.cn id=415 lang=python3
#
# [415] 字符串相加
#

# @lc code=start
from typing import List
class Solution:
    def func(self, n1: List[int], n2: List[int], idx: int, carry: int = 0):
        if carry == 0 and idx >= len(n1) and idx >= len(n2): return
        tmp = (n1[idx] if idx < len(n1) else 0) + (n2[idx] if idx < len(n2) else 0) + carry
        self.ret.append(tmp % 10)
        self.func(n1, n2, idx + 1, tmp // 10)
    def addStrings(self, num1: str, num2: str) -> str:
        n1, n2 = [ord(c) - ord('0') for c in reversed(num1)], [ord(c) - ord('0') for c in reversed(num2)]
        self.ret = []
        self.func(n1, n2, 0)
        return ''.join(map(str, reversed(self.ret)))
# @lc code=end

print(Solution().addStrings(num1 = "11", num2 = "123"))
print(Solution().addStrings(num1 = "456", num2 = "77"))
print(Solution().addStrings(num1 = "0", num2 = "0"))